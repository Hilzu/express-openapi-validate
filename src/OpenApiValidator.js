"use strict";

const Ajv = require("ajv");
const _ = require("lodash");
const semver = require("semver");
const ValidationError = require("./ValidationError");

const validMethods = [
  "get",
  "put",
  "post",
  "delete",
  "options",
  "head",
  "patch",
  "trace",
];

class OpenApiValidator {
  constructor(openApiDocument) {
    if (!semver.satisfies(openApiDocument.openapi, "^3.0.0")) {
      const version = openApiDocument.openapi || openApiDocument.swagger;
      throw new Error(`Unsupported OpenAPI / Swagger version=${version}`);
    }
    this.document = openApiDocument;
    this.ajv = new Ajv();
  }

  getOperationObject(method, path) {
    if (!validMethods.includes(method)) {
      throw new Error(`Not a valid method=${method}`);
    }
    try {
      return this.document.paths[path][method];
    } catch (err) {
      throw new Error(
        `Path=${path} with method=${method} not found from OpenAPI document`
      );
    }
  }

  validate(method, path) {
    const operation = this.getOperationObject(method, path);
    const schema = {
      properties: {
        body: _.get(
          operation,
          ["requestBody", "content", "application/json", "schema"],
          {}
        ),
      },
      required: ["body"],
    };
    const validator = this.ajv.compile(schema);
    const validate = (req, res, next) => {
      const valid = validator(req);
      if (valid) {
        next();
      } else {
        const err = new ValidationError(
          "Error while validating request parameters",
          validator.errors
        );
        next(err);
      }
    };
    return validate;
  }
}

module.exports = OpenApiValidator;
