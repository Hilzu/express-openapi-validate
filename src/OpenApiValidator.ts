/*
  Copyright 2018 Santeri Hiltunen

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

import Ajv, { Options as AjvOptions, ErrorObject } from "ajv";
import addFormats from "ajv-formats";
// eslint-disable-next-line import/no-extraneous-dependencies
import { RequestHandler } from "express";
import _ from "lodash";
import { pathToRegexp } from "path-to-regexp";
import * as semver from "semver";

import debug from "./debug";
import * as formats from "./formats";
import OpenApiDocument, {
  Operation,
  OperationObject,
  PathItemObject,
  SchemaObject,
} from "./OpenApiDocument";
import * as parameters from "./parameters";
import {
  mapOasSchemaToJsonSchema,
  oasPathToExpressPath,
  resolveReference,
} from "./schema-utils";
import ValidationError from "./ValidationError";

const resolveResponse = (res: any): any => {
  if (res == null) {
    throw new TypeError(`Response was ${String(res)}`);
  }
  const statusCodeNum = Number(res.statusCode || res.status);
  const statusCode = Number.isNaN(statusCodeNum) ? null : statusCodeNum;
  const body = res.body || res.data;
  const { headers } = res;
  if (statusCode == null || body == null || headers == null) {
    throw new TypeError(
      "statusCode, body or header values not found from response",
    );
  }
  return { statusCode, body, headers };
};

export interface ValidatorConfig {
  ajvOptions?: AjvOptions;
}

export interface PathRegexpObject {
  path: string;
  regex: RegExp;
}

export interface MatchOptions {
  allowNoMatch?: boolean;
}

export default class OpenApiValidator {
  private _ajv: Ajv;

  private _document: OpenApiDocument;

  constructor(openApiDocument: OpenApiDocument, options: ValidatorConfig = {}) {
    if (!semver.satisfies(openApiDocument.openapi, "^3.0.0")) {
      const version =
        openApiDocument.openapi || (openApiDocument as any).swagger;
      throw new Error(`Unsupported OpenAPI / Swagger version=${version}`);
    }
    this._document = openApiDocument;
    const userAjvFormats = _.get(options, ["ajvOptions", "formats"], {});
    const ajvOptions: AjvOptions = {
      discriminator: true,
      ...options.ajvOptions,
      formats: { ...formats, ...userAjvFormats },
    };
    this._ajv = new Ajv(ajvOptions);
    addFormats(this._ajv, ["date", "date-time"]);
    this._ajv.addKeyword("example");
    this._ajv.addKeyword("xml");
    this._ajv.addKeyword("externalDocs");
  }

  public validate(method: Operation, path: string): RequestHandler {
    const pathItemObject = this._getPathItemObject(path);
    const operation = this._getOperationObject(method, path);
    const requestBodyObject = resolveReference(
      this._document,
      _.get(operation, ["requestBody"], {}),
    );
    const bodySchema = _.get(
      requestBodyObject,
      ["content", "application/json", "schema"],
      {},
    );

    const params = parameters.resolve(
      this._document,
      pathItemObject.parameters,
      operation.parameters,
    );
    const parametersSchema = parameters.buildSchema(params);
    const schema = {
      type: "object",
      properties: {
        body: resolveReference(this._document, bodySchema),
        ...parametersSchema,
      },
      required: ["query", "headers", "params"],
    };
    if (!_.isEmpty(parametersSchema.cookies)) {
      schema.required.push("cookies");
    }

    if (_.get(requestBodyObject, ["required"]) === true) {
      schema.required.push("body");
    }
    const jsonSchema = mapOasSchemaToJsonSchema(schema, this._document);
    const validator = this._ajv.compile(jsonSchema);
    debug(`Request JSON Schema for ${method} ${path}: %j`, jsonSchema);

    const validate: RequestHandler = (req, res, next) => {
      const reqToValidate = {
        ..._.pick(req, "query", "headers", "params", "body"),
        cookies: req.cookies
          ? { ...req.cookies, ...req.signedCookies }
          : undefined,
      };
      const valid = validator(reqToValidate);
      if (valid) {
        next();
      } else {
        const errors = validator.errors as ErrorObject[];
        const errorText = this._ajv.errorsText(errors, { dataVar: "request" });
        const err = new ValidationError(
          `Error while validating request: ${errorText}`,
          errors,
        );
        next(err);
      }
    };

    return validate;
  }

  public match(
    options: MatchOptions = { allowNoMatch: false },
  ): RequestHandler {
    const paths: PathRegexpObject[] = _.keys(this._document.paths).map(
      (path) => ({
        path,
        regex: pathToRegexp(oasPathToExpressPath(path)),
      }),
    );
    const matchAndValidate: RequestHandler = (req, res, next) => {
      const match = paths.find(({ regex }) => regex.test(req.path));
      const method = req.method.toLowerCase() as Operation;
      if (match) {
        this.validate(method, match.path)(req, res, next);
      } else if (!options.allowNoMatch) {
        const err = new Error(
          `Path=${req.path} with method=${method} not found from OpenAPI document`,
        );
        next(err);
      } else {
        // match not required
        next();
      }
    };
    return matchAndValidate;
  }

  public validateResponse(method: Operation, path: string): (res: any) => void {
    const operation = this._getOperationObject(method, path);
    const validateResponse = (userResponse: any): void => {
      const { statusCode, ...response } = resolveResponse(userResponse);
      const responseObject = this._getResponseObject(operation, statusCode);
      const bodySchema = _.get(
        responseObject,
        ["content", "application/json", "schema"],
        {},
      );

      const headerObjectMap = _.get(responseObject, ["headers"], {});
      const headersSchema: SchemaObject = {
        type: "object",
        properties: {},
      };
      Object.keys(headerObjectMap).forEach((key) => {
        const headerObject = resolveReference(
          this._document,
          headerObjectMap[key],
        );
        const name = key.toLowerCase();
        if (name === "content-type") {
          return;
        }
        if (headerObject.required === true) {
          if (!Array.isArray(headersSchema.required)) {
            headersSchema.required = [];
          }
          headersSchema.required.push(name);
        }
        (headersSchema.properties as any)[name] = resolveReference(
          this._document,
          headerObject.schema || {},
        );
      });

      const schema = mapOasSchemaToJsonSchema(
        {
          type: "object",
          properties: {
            body: resolveReference(this._document, bodySchema),
            headers: headersSchema,
          },
          required: ["headers", "body"],
        },
        this._document,
      );

      debug(
        `Response JSON Schema for ${method} ${path} ${statusCode}: %j`,
        schema,
      );

      const valid = this._ajv.validate(schema, response);
      if (!valid) {
        const errorText = this._ajv.errorsText(this._ajv.errors, {
          dataVar: "response",
        });
        throw new ValidationError(
          `Error while validating response: ${errorText}`,
          this._ajv.errors as ErrorObject[],
        );
      }
    };
    return validateResponse;
  }

  private _getResponseObject(op: OperationObject, statusCode: number): any {
    const statusCodeStr = String(statusCode);
    let responseObject = _.get(op, ["responses", statusCodeStr], null);
    if (responseObject === null) {
      const field = `${statusCodeStr[0]}XX`;
      responseObject = _.get(op, ["responses", field], null);
    }
    if (responseObject === null) {
      responseObject = _.get(op, ["responses", "default"], null);
    }
    if (responseObject === null) {
      throw new Error(
        `No response object found with statusCode=${statusCodeStr}`,
      );
    }
    return resolveReference(this._document, responseObject);
  }

  private _getPathItemObject(path: string): PathItemObject {
    if (_.has(this._document, ["paths", path])) {
      return this._document.paths[path] as PathItemObject;
    }
    throw new Error(`Path=${path} not found from OpenAPI document`);
  }

  private _getOperationObject(
    method: Operation,
    path: string,
  ): OperationObject {
    if (_.has(this._document, ["paths", path, method])) {
      return this._document.paths[path][method] as OperationObject;
    }
    throw new Error(
      `Path=${path} with method=${method} not found from OpenAPI document`,
    );
  }
}
