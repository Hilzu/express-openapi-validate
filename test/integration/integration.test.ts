/*
  Copyright 2018 Santeri Hiltunen

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

import request from "supertest";
// eslint-disable-next-line
import OpenApiValidator from "../../dist/OpenApiValidator";
import openApiDocument from "../open-api-document";
import app from "./app";

const validator = new OpenApiValidator(openApiDocument);

describe("Integration tests with real app", () => {
  test("requests against /echo are validated correctly", async () => {
    const validate = validator.validateResponse("post", "/echo");

    let res = await request(app).post("/echo").send({});
    expect(validate(res)).toBeUndefined();
    expect(res.status).toBe(400);
    expect(res.body).toHaveProperty("error");
    expect(res.body).toMatchSnapshot();

    res = await request(app).post("/echo").send({ input: "Hello!" });
    expect(res.status).toBe(200);
    expect(res.body).toEqual({ output: "Hello!" });
    expect(validate(res)).toBeUndefined();
  });

  test("requests against /match are validated correctly", async () => {
    const validate = validator.validateResponse("post", "/match");

    let res = await request(app).post("/match").send({});
    expect(validate(res)).toBeUndefined();
    expect(res.status).toBe(400);
    expect(res.body).toHaveProperty("error");
    expect(res.body).toMatchSnapshot();

    res = await request(app).post("/match").send({ input: "Hello!" });
    expect(res.status).toBe(200);
    expect(res.body).toEqual({ output: "Hello!" });
    expect(validate(res)).toBeUndefined();

    res = await request(app)
      .post("/match/works-with-url-param")
      .send({ input: "Hello!" });
    expect(res.status).toBe(200);
    expect(res.body).toEqual({ output: "works-with-url-param" });
    expect(validate(res)).toBeUndefined();

    res = await request(app).post("/match/works-with-url-param").send({});
    expect(validate(res)).toBeUndefined();
    expect(res.status).toBe(400);
    expect(res.body).toHaveProperty("error");
    expect(res.body).toMatchSnapshot();
  });

  test("requests against /no-match cause an error", async () => {
    const res = await request(app)
      .post("/no-match")
      .send({ anything: "anything" });
    expect(res.status).toBe(500);
    expect(res.body).toHaveProperty("error");
    expect(res.body).toMatchSnapshot();
  });

  test("path parameters are validated", async () => {
    let res = await request(app).get("/parameters/id/lol");
    expect(res.status).toBe(400);
    expect(res.body).toHaveProperty("error");
    expect(res.body).toMatchSnapshot();

    res = await request(app).get("/parameters/id/789");
    expect(res.status).toBe(200);
    expect(res.body).toEqual({ id: 789 });
  });

  test("query parameters are validated", async () => {
    let res = await request(app).get("/parameters");
    expect(res.status).toBe(400);
    expect(res.body).toHaveProperty("error");
    expect(res.body).toMatchSnapshot();

    res = await request(app).get("/parameters").query({ porom: "moi" });
    expect(res.status).toBe(400);
    expect(res.body).toHaveProperty("error");
    expect(res.body).toMatchSnapshot();

    res = await request(app)
      .get("/parameters")
      .query({ param: "hallo", porom: "moi" });
    expect(res.status).toBe(200);
    expect(res.body).toEqual({ param: "hallo", porom: "moi" });
  });

  test("header parameters are validated", async () => {
    let res = await request(app).get("/parameters/header");
    expect(res.status).toBe(400);
    expect(res.body).toHaveProperty("error");
    expect(res.body).toMatchSnapshot();

    res = await request(app).get("/parameters/header").set("X-param", "hullo");
    expect(res.status).toBe(200);
    expect(res.body).toEqual({ header: "hullo" });
  });

  test("cookie parameters are validated", async () => {
    let res = await request(app).get("/parameters/cookie");
    expect(res.status).toBe(400);
    expect(res.body).toHaveProperty("error");
    expect(res.body).toMatchSnapshot();

    res = await request(app)
      .get("/parameters/cookie")
      .set("cookie", "session=hullo");
    expect(res.status).toBe(200);
    expect(res.body).toEqual({ cookie: "hullo" });
  });
});
