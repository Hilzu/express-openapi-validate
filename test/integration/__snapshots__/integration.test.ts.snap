// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests with real app cookie parameters are validated 1`] = `
Object {
  "error": Object {
    "data": Array [
      Object {
        "instancePath": "/cookies",
        "keyword": "required",
        "message": "must have required property 'session'",
        "params": Object {
          "missingProperty": "session",
        },
        "schemaPath": "#/properties/cookies/required",
      },
    ],
    "message": "Error while validating request: request/cookies must have required property 'session'",
    "name": "ValidationError",
  },
}
`;

exports[`Integration tests with real app header parameters are validated 1`] = `
Object {
  "error": Object {
    "data": Array [
      Object {
        "instancePath": "/headers",
        "keyword": "required",
        "message": "must have required property 'x-param'",
        "params": Object {
          "missingProperty": "x-param",
        },
        "schemaPath": "#/properties/headers/required",
      },
    ],
    "message": "Error while validating request: request/headers must have required property 'x-param'",
    "name": "ValidationError",
  },
}
`;

exports[`Integration tests with real app path parameters are validated 1`] = `
Object {
  "error": Object {
    "data": Array [
      Object {
        "instancePath": "/params/id",
        "keyword": "pattern",
        "message": "must match pattern \\"^[0-9]+$\\"",
        "params": Object {
          "pattern": "^[0-9]+$",
        },
        "schemaPath": "#/properties/params/properties/id/pattern",
      },
    ],
    "message": "Error while validating request: request/params/id must match pattern \\"^[0-9]+$\\"",
    "name": "ValidationError",
  },
}
`;

exports[`Integration tests with real app query parameters are validated 1`] = `
Object {
  "error": Object {
    "data": Array [
      Object {
        "instancePath": "/query",
        "keyword": "required",
        "message": "must have required property 'param'",
        "params": Object {
          "missingProperty": "param",
        },
        "schemaPath": "#/properties/query/required",
      },
    ],
    "message": "Error while validating request: request/query must have required property 'param'",
    "name": "ValidationError",
  },
}
`;

exports[`Integration tests with real app query parameters are validated 2`] = `
Object {
  "error": Object {
    "data": Array [
      Object {
        "instancePath": "/query",
        "keyword": "required",
        "message": "must have required property 'param'",
        "params": Object {
          "missingProperty": "param",
        },
        "schemaPath": "#/properties/query/required",
      },
    ],
    "message": "Error while validating request: request/query must have required property 'param'",
    "name": "ValidationError",
  },
}
`;

exports[`Integration tests with real app requests against /echo are validated correctly 1`] = `
Object {
  "error": Object {
    "data": Array [
      Object {
        "instancePath": "/body",
        "keyword": "required",
        "message": "must have required property 'input'",
        "params": Object {
          "missingProperty": "input",
        },
        "schemaPath": "#/properties/body/required",
      },
    ],
    "message": "Error while validating request: request/body must have required property 'input'",
    "name": "ValidationError",
  },
}
`;

exports[`Integration tests with real app requests against /match are validated correctly 1`] = `
Object {
  "error": Object {
    "data": Array [
      Object {
        "instancePath": "/body",
        "keyword": "required",
        "message": "must have required property 'input'",
        "params": Object {
          "missingProperty": "input",
        },
        "schemaPath": "#/properties/body/required",
      },
    ],
    "message": "Error while validating request: request/body must have required property 'input'",
    "name": "ValidationError",
  },
}
`;

exports[`Integration tests with real app requests against /match are validated correctly 2`] = `
Object {
  "error": Object {
    "data": Array [
      Object {
        "instancePath": "/body",
        "keyword": "required",
        "message": "must have required property 'input'",
        "params": Object {
          "missingProperty": "input",
        },
        "schemaPath": "#/properties/body/required",
      },
    ],
    "message": "Error while validating request: request/body must have required property 'input'",
    "name": "ValidationError",
  },
}
`;

exports[`Integration tests with real app requests against /no-match cause an error 1`] = `
Object {
  "error": Object {
    "message": "Path=/no-match with method=post not found from OpenAPI document",
    "name": "Error",
  },
}
`;
